File: C++/BucketSort.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

void bucketSort(float arr[], int n)
{
    vector<float> *b = new vector<float>[n]; // create N empty buckets

    for (int i = 0; i < n; i++) // put array element in different bucket
    {
        int bi = n * arr[i];
        b[bi].push_back(arr[i]);
    }

    for (int i = 0; i < n; i++) // sort individual bucket
        sort(b[i].begin(), b[i].end());

    int index = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < b[i].size(); j++)
            arr[index++] = b[i][j];

    // delete[] b;
}

int main(void)
{
    int n;
    cout << "\nEnter the number of elements: ";
    cin >> n;

    float arr[n];
    cout << "\nEnter " << n << " elements (<1 and >0):\n";
    for (int i = 0; i < n; i++)
        cin >> arr[i];

    bucketSort(arr, n);

    cout << "\nSorted array:";
    for (int i = 0; i < n; i++)
        cout << " " << arr[i];
}

/*
Enter the number of elements: 5
Enter 5 elements:
0.7 0.9 0.4 0.28 0.6
Sorted array: 0.28 0.4 0.6 0.7 0.9
*/

File: Python/BucketSort.py
def insertion(b):
    for i in range(1, len(b)):
        up = b[i]
        j = i-1
        while j >= 0 and b[j] > up:
            b[j+1] = b[j]
            j -= 1
        b[j+1] = up
    return b


def bucketSort(x):
    arr = []
    slot = 10  # 10 means 10 slots, each slot's size is 0.1
    for i in range(slot):
        arr.append([])

    for j in x:  # put array element in different bucket
        index = int(slot*j)
        arr[index].append(j)

    for i in range(slot):  # sort individual bucket
        arr[i] = insertion(arr[i])

    k = 0
    for i in range(slot):  # concatenate the results
        for j in range(len(arr[i])):
            x[k] = arr[i][j]
            k += 1

    return x


x = [0.84, 0.68, 0.71, 0.28, 0.59]

print(bucketSort(x))
